File: src/source/meta.scala
/**
  * Copyright 2014 Dropbox, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *    http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  * 
  * This file has been modified by Snap, Inc.
  */

package djinni

import djinni.ast.TypeDef
import djinni.ast.ProtobufMessage
import scala.collection.immutable

package object meta {

case class MExpr(base: Meta, args: Seq[MExpr])

abstract sealed class Meta
{
  val numParams: Int
}

case class MParam(name: String) extends Meta { val numParams = 0 }
case class MDef(name: String, override val numParams: Int, defType: DefType, body: TypeDef) extends Meta
case class MExtern(name: String, override val numParams: Int, defType: DefType, body: TypeDef, cpp: MExtern.Cpp, objc: MExtern.Objc, objcpp: MExtern.Objcpp, java: MExtern.Java, jni: MExtern.Jni, wasm: MExtern.Wasm, ts: MExtern.Ts) extends Meta
object MExtern {
  // These hold the information marshals need to interface with existing types correctly
  // All include paths are complete including quotation marks "a/b/c" or angle brackets <a/b/c>.
  // All typenames are fully qualified in their respective language.
  // TODO: names of enum values and record fields as written in code for use in constants (do not use IdentStyle)
  case class Cpp(
    typename: String,
    header: String,
    byValue: Boolean, // Whether to pass struct by value in C++ (e.g. std::chrono::duration). Only used for "record" types.
    moveOnly: Boolean
  )
  case class Objc(
    typename: String,
    header: String,
    boxed: String, // Fully qualified Objective-C typename, must be an object. Only used for "record" types.
    pointer: Boolean, // True to construct pointer types and make it eligible for "nonnull" qualifier. Only used for "record" types.
    generic: Boolean, // Set to false to exclude type arguments from the ObjC class. This is should be true by default. Useful if template arguments are only used in C++.
    hash: String, // A well-formed expression to get the hash value. Must be a format string with a single "%s" placeholder. Only used for "record".
    equal: String, // Set equal operator. E.i: .isEqual: for default operator
    protocol: Boolean
  )
  case class Objcpp(
    translator: String, // C++ typename containing toCpp/fromCpp methods
    header: String // Where to find the translator class
  )
  case class Java(
    typename: String,
    boxed: String, // Java typename used if boxing is required, must be an object.
    reference: Boolean, // True if the unboxed type is an object reference and qualifies for any kind of "nonnull" annotation in Java. Only used for "record" types.
    generic: Boolean, // Set to false to exclude type arguments from the Java class. This is should be true by default. Useful if template arguments are only used in C++.
    hash: String, // A well-formed expression to get the hash value. Must be a format string with a single "%s" placeholder. Only used for "record" types.
    writeToParcel: String, // A well-formed expression to write value into android.os.Parcel. Must be a format string with a single "%s" placeholder. Only used for "record" types types
    readFromParcel: String // A well-formed expression to read value from android.os.Parcel. Must be a format string with a single "%s" placeholder. Only used for "record" types types
  )
  case class Jni(
    translator: String, // C++ typename containing toCpp/fromCpp methods
    header: String, // Where to find the translator class
    typename: String, // The JNI type to use (e.g. jobject, jstring)
    typeSignature: String // The mangled Java type signature (e.g. "Ljava/lang/String;")
  )
  case class Wasm(
    typename: String, // The Emscripten type to use (e.g. em::val, int32_t)
    translator: String, // C++ typename containing toCpp/fromCpp methods
    header: String // Where to find the translator class
  )
  case class Ts(
    typename: String, // The TypeScript type
    module: String,   // The module to import for the type
    generic: Boolean
  )
}
case class MProtobuf(name: String, override val numParams: Int, body: ProtobufMessage) extends Meta

abstract sealed class MOpaque extends Meta { val idlName: String }

abstract sealed class DefType
case object DEnum extends DefType
case object DInterface extends DefType
case object DRecord extends DefType

case class MPrimitive(_idlName: String, jName: String, jniName: String, cName: String, jBoxed: String, jSig: String, objcName: String, objcBoxed: String, kName: String) extends MOpaque { val numParams = 0; val idlName = _idlName }
case object MString extends MOpaque { val numParams = 0; val idlName = "string" }
case object MDate extends MOpaque { val numParams = 0; val idlName = "date" }
case object MBinary extends MOpaque { val numParams = 0; val idlName = "binary" }
case object MOptional extends MOpaque { val numParams = 1; val idlName = "optional" }
case object MList extends MOpaque { val numParams = 1; val idlName = "list" }
case object MSet extends MOpaque { val numParams = 1; val idlName = "set" }
case object MMap extends MOpaque { val numParams = 2; val idlName = "map" }
case object MArray extends MOpaque { val numParams = 1; val idlName = "array"}
case object MVoid extends MOpaque { val numParams = 0; val idlName = "void"}

val defaults: Map[String,MOpaque] = immutable.HashMap(
  ("i8",   MPrimitive("i8",   "byte",    "jbyte",    "int8_t",  "Byte",    "B", "int8_t",  "NSNumber", "Byte")),
  ("i16",  MPrimitive("i16",  "short",   "jshort",   "int16_t", "Short",   "S", "int16_t", "NSNumber", "Short")),
  ("i32",  MPrimitive("i32",  "int",     "jint",     "int32_t", "Integer", "I", "int32_t", "NSNumber", "Int")),
  ("i64",  MPrimitive("i64",  "long",    "jlong",    "int64_t", "Long",    "J", "int64_t", "NSNumber", "Long")),
  ("f32",  MPrimitive("f32",  "float",   "jfloat",   "float",   "Float",   "F", "float",   "NSNumber", "Float")),
  ("f64",  MPrimitive("f64",  "double",  "jdouble",  "double",  "Double",  "D", "double",  "NSNumber", "Double")),
  ("bool", MPrimitive("bool", "boolean", "jboolean", "bool",    "Boolean", "Z", "BOOL",    "NSNumber", "Boolean")),
  ("string", MString),
  ("binary", MBinary),
  ("optional", MOptional),
  ("date", MDate),
  ("list", MList),
  ("set", MSet),
  ("map", MMap),
  ("array", MArray),
  ("void", MVoid))

def isInterface(ty: MExpr): Boolean = {
  ty.base match {
    case d: MDef => d.defType == DInterface
    case e: MExtern => e.defType == DInterface
    case _ => false
  }
}

def isOptional(ty: MExpr): Boolean = {
  ty.base == MOptional && ty.args.length == 1
}

def isOptionalInterface(ty: MExpr): Boolean = {
  isOptional(ty) && isInterface(ty.args.head)
}
}

File: djinni-rs/crates/djinni-ast/src/meta.rs
// vinsg
// Meta type system for resolved types, translated from meta.scala

use std::collections::HashMap;

use crate::ast::{ProtobufMessage, TypeDef};

#[derive(Debug, Clone)]
pub struct MExpr {
    pub base: Meta,
    pub args: Vec<MExpr>,
}

#[derive(Debug, Clone)]
pub enum Meta {
    MParam(MParam),
    MDef(MDef),
    MExtern(Box<MExtern>),
    MProtobuf(MProtobuf),
    MPrimitive(MPrimitive),
    MString,
    MDate,
    MBinary,
    MOptional,
    MList,
    MSet,
    MMap,
    MArray,
    MVoid,
}

impl Meta {
    pub fn num_params(&self) -> usize {
        match self {
            Meta::MParam(_) => 0,
            Meta::MDef(d) => d.num_params,
            Meta::MExtern(e) => e.num_params,
            Meta::MProtobuf(_) => 0,
            Meta::MPrimitive(_) => 0,
            Meta::MString | Meta::MDate | Meta::MBinary | Meta::MVoid => 0,
            Meta::MOptional | Meta::MList | Meta::MSet | Meta::MArray => 1,
            Meta::MMap => 2,
        }
    }

    pub fn idl_name(&self) -> &str {
        match self {
            Meta::MPrimitive(p) => &p.idl_name,
            Meta::MString => "string",
            Meta::MDate => "date",
            Meta::MBinary => "binary",
            Meta::MOptional => "optional",
            Meta::MList => "list",
            Meta::MSet => "set",
            Meta::MMap => "map",
            Meta::MArray => "array",
            Meta::MVoid => "void",
            _ => "",
        }
    }
}

#[derive(Debug, Clone)]
pub struct MParam {
    pub name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DefType {
    Enum,
    Interface,
    Record,
}

#[derive(Debug, Clone)]
pub struct MDef {
    pub name: String,
    pub num_params: usize,
    pub def_type: DefType,
    pub body: TypeDef,
}

#[derive(Debug, Clone)]
pub struct MExtern {
    pub name: String,
    pub num_params: usize,
    pub def_type: DefType,
    pub body: TypeDef,
    pub cpp: ExternCpp,
    pub objc: ExternObjc,
    pub objcpp: ExternObjcpp,
    pub java: ExternJava,
    pub jni: ExternJni,
    pub wasm: ExternWasm,
    pub ts: ExternTs,
}

#[derive(Debug, Clone)]
pub struct ExternCpp {
    pub typename: String,
    pub header: String,
    pub by_value: bool,
    pub move_only: bool,
}

#[derive(Debug, Clone)]
pub struct ExternObjc {
    pub typename: String,
    pub header: String,
    pub boxed: String,
    pub pointer: bool,
    pub generic: bool,
    pub hash: String,
    pub equal: String,
    pub protocol: bool,
}

#[derive(Debug, Clone)]
pub struct ExternObjcpp {
    pub translator: String,
    pub header: String,
}

#[derive(Debug, Clone)]
pub struct ExternJava {
    pub typename: String,
    pub boxed: String,
    pub reference: bool,
    pub generic: bool,
    pub hash: String,
    pub write_to_parcel: String,
    pub read_from_parcel: String,
}

#[derive(Debug, Clone)]
pub struct ExternJni {
    pub translator: String,
    pub header: String,
    pub typename: String,
    pub type_signature: String,
}

#[derive(Debug, Clone)]
pub struct ExternWasm {
    pub typename: String,
    pub translator: String,
    pub header: String,
}

#[derive(Debug, Clone)]
pub struct ExternTs {
    pub typename: String,
    pub module: String,
    pub generic: bool,
}

#[derive(Debug, Clone)]
pub struct MProtobuf {
    pub name: String,
    pub num_params: usize,
    pub body: ProtobufMessage,
}

#[derive(Debug, Clone)]
pub struct MPrimitive {
    pub idl_name: String,
    pub j_name: String,
    pub jni_name: String,
    pub c_name: String,
    pub j_boxed: String,
    pub j_sig: String,
    pub objc_name: String,
    pub objc_boxed: String,
    pub k_name: String,
}

pub fn is_interface(ty: &MExpr) -> bool {
    match &ty.base {
        Meta::MDef(d) => d.def_type == DefType::Interface,
        Meta::MExtern(e) => e.def_type == DefType::Interface,
        _ => false,
    }
}

pub fn is_optional(ty: &MExpr) -> bool {
    matches!(&ty.base, Meta::MOptional) && ty.args.len() == 1
}

pub fn is_optional_interface(ty: &MExpr) -> bool {
    is_optional(ty) && is_interface(&ty.args[0])
}

pub fn defaults() -> HashMap<String, Meta> {
    let mut m = HashMap::new();
    m.insert("i8".into(), Meta::MPrimitive(MPrimitive {
        idl_name: "i8".into(), j_name: "byte".into(), jni_name: "jbyte".into(),
        c_name: "int8_t".into(), j_boxed: "Byte".into(), j_sig: "B".into(),
        objc_name: "int8_t".into(), objc_boxed: "NSNumber".into(), k_name: "Byte".into(),
    }));
    m.insert("i16".into(), Meta::MPrimitive(MPrimitive {
        idl_name: "i16".into(), j_name: "short".into(), jni_name: "jshort".into(),
        c_name: "int16_t".into(), j_boxed: "Short".into(), j_sig: "S".into(),
        objc_name: "int16_t".into(), objc_boxed: "NSNumber".into(), k_name: "Short".into(),
    }));
    m.insert("i32".into(), Meta::MPrimitive(MPrimitive {
        idl_name: "i32".into(), j_name: "int".into(), jni_name: "jint".into(),
        c_name: "int32_t".into(), j_boxed: "Integer".into(), j_sig: "I".into(),
        objc_name: "int32_t".into(), objc_boxed: "NSNumber".into(), k_name: "Int".into(),
    }));
    m.insert("i64".into(), Meta::MPrimitive(MPrimitive {
        idl_name: "i64".into(), j_name: "long".into(), jni_name: "jlong".into(),
        c_name: "int64_t".into(), j_boxed: "Long".into(), j_sig: "J".into(),
        objc_name: "int64_t".into(), objc_boxed: "NSNumber".into(), k_name: "Long".into(),
    }));
    m.insert("f32".into(), Meta::MPrimitive(MPrimitive {
        idl_name: "f32".into(), j_name: "float".into(), jni_name: "jfloat".into(),
        c_name: "float".into(), j_boxed: "Float".into(), j_sig: "F".into(),
        objc_name: "float".into(), objc_boxed: "NSNumber".into(), k_name: "Float".into(),
    }));
    m.insert("f64".into(), Meta::MPrimitive(MPrimitive {
        idl_name: "f64".into(), j_name: "double".into(), jni_name: "jdouble".into(),
        c_name: "double".into(), j_boxed: "Double".into(), j_sig: "D".into(),
        objc_name: "double".into(), objc_boxed: "NSNumber".into(), k_name: "Double".into(),
    }));
    m.insert("bool".into(), Meta::MPrimitive(MPrimitive {
        idl_name: "bool".into(), j_name: "boolean".into(), jni_name: "jboolean".into(),
        c_name: "bool".into(), j_boxed: "Boolean".into(), j_sig: "Z".into(),
        objc_name: "BOOL".into(), objc_boxed: "NSNumber".into(), k_name: "Boolean".into(),
    }));
    m.insert("string".into(), Meta::MString);
    m.insert("binary".into(), Meta::MBinary);
    m.insert("optional".into(), Meta::MOptional);
    m.insert("date".into(), Meta::MDate);
    m.insert("list".into(), Meta::MList);
    m.insert("set".into(), Meta::MSet);
    m.insert("map".into(), Meta::MMap);
    m.insert("array".into(), Meta::MArray);
    m.insert("void".into(), Meta::MVoid);
    m
}
